File: Back-End/src/main/java/iuh/fit/ecommerce/services/PromotionService.java
package iuh.fit.ecommerce.services;

import iuh.fit.ecommerce.dtos.request.promotion.PromotionAddRequest;
import iuh.fit.ecommerce.dtos.request.promotion.PromotionUpdateRequest;
import iuh.fit.ecommerce.dtos.response.base.ResponseWithPagination;
import iuh.fit.ecommerce.dtos.response.promotion.PromotionResponse;
import iuh.fit.ecommerce.entities.Product;
import iuh.fit.ecommerce.entities.ProductVariant;
import iuh.fit.ecommerce.entities.Promotion;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public interface PromotionService {
    PromotionResponse createPromotion(PromotionAddRequest request);
    PromotionResponse getPromotionById(Long id);
    ResponseWithPagination<List<PromotionResponse>> getAllPromotions(
            int page, int limit,
            String name, String type,
            Boolean active,
            LocalDate startDate,
            Integer priority
    );
    PromotionResponse updatePromotion(Long id, PromotionUpdateRequest request);
    void deletePromotion(Long id);
    void changeStatusPromotion(Long id);
    Double calculateDiscountPrice(ProductVariant variant, Promotion promotion);
    Double calculateOriginalPrice(ProductVariant variant);
    Promotion getBestPromotion(ProductVariant variant, Map<Long, List<Promotion>> promosByVariant);

    Map<Long, List<Promotion>>  getPromotionsGroupByVariantId(List<ProductVariant> variants, Product product);

    Promotion getBestPromotionForVariant(ProductVariant variant);
}

File: Front-End/src/services/promotion.service.ts
import axiosClient from '@/configurations/axios.config';
import type {
  CreatePromotionRequest,
  UpdatePromotionRequest,
  PromotionListResponse,
  PromotionResponse,
} from "@/types/promotion.type";

interface GetPromotionsParams {
  page: number;
  size: number;
  name?: string;
  type?: string;
  active?: boolean;
  startDate?: string;
  endDate?: string;
}

export const promotionService = {
  getPromotions: async ({
    page,
    size,
    name,
    type,
    active,
    startDate,
    endDate,
  }: GetPromotionsParams) => {
    const params = new URLSearchParams({
      page: page.toString(),
      limit: size.toString(),
    });

    if (name) params.append("name", name);
    if (type) params.append("type", type);
    if (active !== undefined) params.append("active", active.toString());
    if (startDate) params.append("startDate", startDate);
    if (endDate) params.append("endDate", endDate);

    const response = await axiosClient.get<PromotionListResponse>(
      `/promotions?${params.toString()}`
    );
    return response.data;
  },

  getPromotionById: async (id: number) => {
    const response = await axiosClient.get<PromotionResponse>(`/promotions/${id}`);
    return response.data;
  },

  createPromotion: async (request: CreatePromotionRequest) => {
    const response = await axiosClient.post<PromotionResponse>('/promotions', request);
    return response.data;
  },

  updatePromotion: async (id: number, data: UpdatePromotionRequest) => {
    const response = await axiosClient.put<PromotionResponse>(`/promotions/${id}`, data);
    return response.data;
  },

  deletePromotion: async (id: number) => {
    await axiosClient.delete(`/promotions/${id}`);
  },

  changeStatusPromotion: async (id: number) => {
    await axiosClient.put(`/promotions/change-status/${id}`);
  }
};
